backend/msgpack/topology_options.go

```diff
@@ -0,0 +1,129 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package msgpack
+
+import (
+	"time"
+
+	"github.com/m3db/m3cluster/kv"
+
+	"github.com/spaolacci/murmur3"
+)
+
+const (
+	defaultTopologyKey      = "/topology"
+	defaultInitWatchTimeout = 10 * time.Second
+)
+
+// HashGenFn generates hash functions based on the number of shards.
+type HashGenFn func(numShards int) HashFn
+
+// HashFn hashes a metric id to a shard.
+type HashFn func(id []byte) uint32
+
+// TopologyOptions provide a set of topology options.
+type TopologyOptions interface {
+	// SetTopologyKey sets the kv key to watch for topology changes.
+	SetTopologyKey(value string) TopologyOptions
+
+	// TopologyKey returns the kv key to watch for topology changes.
+	TopologyKey() string
+
+	// SetInitWatchTimeout sets the initial watch timeout.
+	SetInitWatchTimeout(value time.Duration) TopologyOptions
+
+	// InitWatchTimeout returns the initial watch timeout.
+	InitWatchTimeout() time.Duration
+
+	// SetKVStore sets the kv store.
+	SetKVStore(store kv.Store) TopologyOptions
+
+	// KVStore returns the kv store.
+	KVStore() kv.Store
+
+	// SetHashGenFn sets the hash generating function.
+	SetHashGenFn(value HashGenFn) TopologyOptions
+
+	// HashGenFn returns the hash generating function.
+	HashGenFn() HashGenFn
+}
+
+type topologyOptions struct {
+	topologyKey      string
+	initWatchTimeout time.Duration
+	kvStore          kv.Store
+	hashGenFn        HashGenFn
+}
+
+// NewTopologyOptions create a new set of topology options.
+func NewTopologyOptions() TopologyOptions {
+	return &topologyOptions{
+		topologyKey:      defaultTopologyKey,
+		initWatchTimeout: defaultInitWatchTimeout,
+		hashGenFn:        defaultHashGen,
+	}
+}
+
+func (o *topologyOptions) SetTopologyKey(value string) TopologyOptions {
+	opts := *o
+	opts.topologyKey = value
+	return &opts
+}
+
+func (o *topologyOptions) TopologyKey() string {
+	return o.topologyKey
+}
+
+func (o *topologyOptions) SetInitWatchTimeout(value time.Duration) TopologyOptions {
+	opts := *o
+	opts.initWatchTimeout = value
+	return &opts
+}
+
+func (o *topologyOptions) InitWatchTimeout() time.Duration {
+	return o.initWatchTimeout
+}
+
+func (o *topologyOptions) SetKVStore(value kv.Store) TopologyOptions {
+	opts := *o
+	opts.kvStore = value
+	return &opts
+}
+
+func (o *topologyOptions) KVStore() kv.Store {
+	return o.kvStore
+}
+
+func (o *topologyOptions) SetHashGenFn(value HashGenFn) TopologyOptions {
+	opts := *o
+	opts.hashGenFn = value
+	return &opts
+}
+
+func (o *topologyOptions) HashGenFn() HashGenFn {
+	return o.hashGenFn
+}
+
+func defaultHashGen(numShards int) HashFn {
```

Comment:
nit: perhaps we want to use consistent hashing here to minimize churn when we add new capacity to the agg tier? We could use [jump hash](https://github.com/m3db/m3x/tree/master/hash/jump) so it would be something like:

```
return jump.Hash(uint64(murmur3.Sum32), numShards)
```

super nit:

I did some [benchmarks on different non-cryptographic hash functions](https://github.com/jeromefroe/golang_benchmarks#non-cryptographic-hash-functions) a while back and we could probably save some cycles by using a hash function thats optimized for 64 bit CPUs like [FarmHash](https://github.com/dgryski/go-farm) or [CityHash](https://github.com/zhenjl/cityhash).